### ISSUE TITLE
provide editing tools for usfm projects (bible)

### ISSUE THREAD

**Issue by neutrinog** [unfoldingWord-dev/ts-desktop#251]
https://github.com/unfoldingWord-dev/ts-requirements/wiki/USFM-Editing-Features

**Comment by soccerhand** [comment]
Cool, I'm out of town for the weekend but will get on it after that.

**Comment by soccerhand** [comment]
Or maybe I'm unassigned now... at any rate I'll check in after the weekend.

**Comment by neutrinog** [comment]
I couldn't make up my mind :)
Let's talk before getting started on this. We still need to figure out how we will do the ui.

**Comment by neutrinog** [comment]
from nathan:
```
Ok I dug into a bit and before I go any further let me make sure I have this process straight.

1. Users translate stuff, adding chapter/verse markers, paragraphs, quotes, etc (via a TBD UI)
2. This text is saved, including the Format Markers, eg:
\c 1
\s1 The Preaching of John the Baptist
\r (Matthew 3.1-12; Luke 3.1-18; John 1.19-28)
\p
\v 1 This is the Good News about Jesus Christ, the Son of God.
\v 2 It began as the prophet Isaiah had written:
\q1 “God said, ‘I will send my messenger ahead of you
\q2 to open the way for you.’

3. The text with markers then needs to be able to be rendered in an appropriate manner (eg: verse markers using superscript, chapter marks as h1's perhaps)

And #3 is the task I am assigned, correct? With emphasis on parsing the Markers and providing hooks for rendering them, rather than what the actual rendered format for each marker should be)


I see this repo out there: https://github.com/unfoldingWord-dev/USFM-Tools
(which is forked from an old copy of https://github.com/openenglishbible/USFM-Tools/)
The readme on this suggests using it as a batch process to convert files into USX format (not sure how that would help us). But should I understand that this is outdated information? The current plan is to render USFM marked up text in realtime?
```

@soccerhand You are correct on `#3` . Also correct that usfm tools does not apply. We'll want to mark the text up in real time. For an example check out https://stackedit.io/editor#!provider=couchdb&id=DW3eUJCh2MWBuZEiIzJk9vB6. This might actually be the tool we'll use for markdown projects (obs for now) though I'm not sure if we will continue to display the markdown syntax inline.

ideally this solution will be in the form of an npm module that can be used to add this enhancement to some view.

**Comment by neutrinog** [comment]
@jag3773 do we need to completely hide the USFM syntax? If we didn't have to this could be greatly simplified since we could basically fork stackedit and make a USFM version.

**Comment by neutrinog** [comment]
@soccerhand this is only loosly related and for a completely different platform but it might be mildly informative. https://github.com/unfoldingWord-dev/Dokuwiki-USFMTag

**Comment by jag3773** [comment]
@neutrinog I think it would be acceptable to take the StackEdit approach.  This would allow advanced users to insert whatever sort of USFM they want, without us having to build a UI for each type.  If we go this way I think we would want a "preview" mode so that they could get a visual on how it looks without the markup syntax, too.  Let me float this past @benjore and @timjore ...

**Comment by jag3773** [comment]
@neutrinog This probably makes it harder, but what if the default was WYSIWYG but there is a toggle to enable advanced mode where you see and can edit the markup too?  I think that's the best of both worlds from the user's standpoint, probably the worst of both worlds from the dev standpoint :(

**Comment by neutrinog** [comment]
perhaps we can start with the advanced mode (since that is basically already available) and add in the WYSIWYG mode later?

**Comment by soccerhand** [comment]
I looked into StackEdit and the actual conversion of markdown to HTML is just done by one file (Markdown.Converter.js). The rest of it is the GUI and other features (including a side-by-side preview). @neutrinog were you suggesting a full replacement of the editor in ts-desktop with stackedit? Or really just getting at the markdown to html conversion piece?
Also looking at the php project you mentioned (https://github.com/unfoldingWord-dev/Dokuwiki-USFMTag) - I think it would be easiest to port this php code to js. (This one looks easier to port than the python at https://github.com/openenglishbible/USFM-Tools/).
I'll start by focusing on creating a js converter for usfm markdown to html.

**Comment by neutrinog** [comment]
@soccerhand yes for now just the USFM to html. We can have one of the Front end devs put the UI for it into a new Polymer element later.

Since we're basically going the route of StackEdit I'm not sure if we want to create an actual fork of their project or just create a custom version of their `Markdown.Converter.js`. Is that file modular? If so perhaps we could just create a fork and add a new converter called `USFM.Converter.js`?

**Comment by soccerhand** [comment]
@neutrinog Turns out the markdown conversion happens in a package called [pagedown](https://github.com/ujifgc/pagedown). It includes the Converter and the Editor (with realtime preview). I'll take a fork of that for now and add the USFM converter. If later we decide to actually fork stackedit, it should be easy enough to just require the forked pagedown package.
It looks like there was not a modular intent in stackedit and pagedown - it's all pretty much focused on markdown syntax. So it strays a little from the intent to add another converter (USFM) but I think it'll work out. If not, no problem, we'll still have our USFM converter.

**Comment by soccerhand** [comment]
@neutrinog I got a port of the php conversion code to run. There's some oddities about the output, and it still definitely needs more work and testing, but it's progress.
https://github.com/soccerhand/pagedownUSFM if you're interested

Is there a place where I can see the [DokuWiki plugin](https://github.com/unfoldingWord-dev/Dokuwiki-USFMTag) (the php code) in action so I can confirm if the oddities I'm seeing are an artifact of my work or are even possibly expected behavior? Thanks.

**Comment by neutrinog** [comment]
@soccerhand here's an example https://door43.org/en/ulb/ep/01-gen.usfm

**Comment by soccerhand** [comment]
@neutrinog we're making progress. Here's a side-by-side of some content
from DokuWiki and the new awesome js decoder. I'll work on more testing
next.
[image: Inline image 1]

**Comment by soccerhand** [comment]
@neutrinog we're making progress. Here's a side-by-side of some content
from DokuWiki and the new awesome js decoder. I'll work on more testing
next.
![image](https://cloud.githubusercontent.com/assets/12506337/13169685/2c1892ce-d69c-11e5-87a3-fda903335096.png)

**Comment by neutrinog** [comment]
great! Can this stand alone from stackedit?

**Comment by soccerhand** [comment]
Yeah it's a totally standalone module. require it, new instance, then let
html = decoder.decode(input). something like that:)

**Comment by soccerhand** [comment]
@neutrinog I should note that this is using [phpjs](https://www.npmjs.com/package/phpjs), as that was the quickest way to get to a working state. Let me know if you'd like me to do anything towards removing this dependency.

**Comment by soccerhand** [comment]
working on integrating the module into ts-desktop: https://github.com/soccerhand/ts-desktop/tree/feature/usfmDecoder

**Comment by neutrinog** [comment]
@soccerhand could you coordinate with @tjklemz on integrating this? This is closely related to markdown support which I think we need to have first.

**Comment by RRohrssen** [comment]
This is related to efforts in #354.

**Comment by neutrinog** [comment]
@soccerhand I ran the test and it cuts out a few lines and leaves a `-.` in their place.

```
   -.
   -.
<p class='usfm-flush' align='justify'><span class='usfm-c'><big class='usfm-c'><big class='usfm-c'><big class='usfm-c'><big class='usfm-c'>1</big></big></big></big></span> <span class='usfm-v'><b class='usfm'><a name='1_1'></a>1</b></span>Blessed is the man who does not walk in the advice of the wicked,</p>
<blockquote class='usfm'>
<blockquote class='usfm'>
<p class='usfm-hanging' align='left'>or stand in the pathway with sinners, </p>
   -.
 <span class='usfm-v'><b class='usfm'><a name='1_2'></a>2</b></span>But his delight is in the law of Yahweh,</p>
   -.
 <span class='usfm-v'><b class='usfm'><a name='1_3'></a>3</b></span>He will be like a tree planted by the streams of water</p>
<p class='usfm-hanging' align='left'>that produces its fruit in its season, </p>
<p class='usfm-hanging' align='left'>whose leaves do not wither; </p>
<p class='usfm-hanging' align='left'>whatever he does will prosper. </p>
</blockquote>
</blockquote>
<hr><h4 class='usfm'></h4>
```

**Comment by neutrinog** [comment]
@soccerhand fyi. I mostly finished this issue just before the baby came. The next step will be able to convert html to usfm. Feel free to begin some research into that.

**Comment by neutrinog** [comment]
@RRohrssen we may want to split this into two issues. USFM to HTML converter and HTML to USFM converter.

**Comment by neutrinog** [comment]
This particular issue is mostly done. I need to finish writing up some tests for it and publish it as an npm module.
